bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -0,0 +1,407 @@
+    <*> O.switch (O.long "check" <> O.help "Check generated key distribution")
+    <*> O.option O.auto (O.long "seed" <> O.value 1337 <> O.showDefault <> O.help "Random seed")
+
+-------------------------------------------------------------------------------
+-- clock
+-------------------------------------------------------------------------------
+
+timed :: IO a -> IO (a, Double)
+timed action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    x  <- action
+    t2 <- Clock.getTime Clock.Monotonic
+    let !t = fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+    return (x, t)
+
+timed_ :: IO () -> IO Double
+timed_ action = do
+    t1 <- Clock.getTime Clock.Monotonic
+    action
+    t2 <- Clock.getTime Clock.Monotonic
+    return $! fromIntegral (Clock.toNanoSecs (Clock.diffTimeSpec t2 t1)) * 1e-9
+
+-------------------------------------------------------------------------------
+-- setup
+-------------------------------------------------------------------------------
+
+-- https://input-output-hk.github.io/fs-sim
+doSetup :: GlobalOpts -> SetupOpts -> IO ()
+doSetup gopts _opts = do
+    let mountPoint :: FS.MountPoint
+        mountPoint = FS.MountPoint gopts.rootDir
+
+    let someFs :: FS.SomeHasFS IO
+        someFs = FS.SomeHasFS (FsIO.ioHasFS mountPoint)
+
+    name <- maybe (fail "invalid snapshot name") return $
+        LSM.mkSnapshotName "bench"
+
+    withSession someFs (FS.mkFsPath []) $ \session -> do
+        tbh <- LSM.new @IO @K @V @B session defaultTableConfig
+
+        forM_ [ 0 .. gopts.initialSize ] $ \ (fromIntegral -> i) -> do
+            -- TODO: this procedure simply inserts all the keys into initial lsm tree
+            -- We might want to do deletes, so there would be delete-insert pairs
+            -- Let's do that when we can actually test that benchmark works.
+
+            let k = makeKey i
+            let v = theValue
+
+            -- TODO: LSM.inserts has annoying order
+            flip LSM.inserts tbh
+                [ (k, v, Nothing)
+                ]
+
+        LSM.snapshot name tbh
+
+-------------------------------------------------------------------------------
+-- dry-run
+-------------------------------------------------------------------------------
+
+doDryRun :: GlobalOpts -> RunOpts -> IO ()
+doDryRun gopts opts = do
+    time <- timed_ $ doDryRun' gopts opts
+    printf "Batch generation: %.03f sec\n" time
+
+doDryRun' :: GlobalOpts -> RunOpts -> IO ()
+doDryRun' gopts opts = do
+    -- calculated some expected statistics for generated batches
+    id $ do
+        -- we generate n random numbers in range of [ 1 .. d ]
+        -- what is the chance they are all distinct
+        let n = fromIntegral (opts.batchCount * opts.batchSize) :: Double
+        let d = fromIntegral gopts.initialSize :: Double
+        -- this is birthday problem.
+        let p = 1 - exp (negate $  (n * (n - 1)) / (2 * d))
```

Comment:
Does this account for the keys that are inserted/deleted during the run?

Reply:
You mean in we don't just have `d = initialSize` "days", but the number of possible keys increases over the course of a run? It doesn't account for that (so it overestimates the expected collisions), but as long as we have relatively few batches, it doesn't make much of a difference.

Reply:
Yup, that's what I was referring to -- I also came to the conclusion you mentioned, but I wanted someone else to verify